crypto/evp/pmeth_lib.c

```diff
@@ -199,6 +199,7 @@ static EVP_PKEY_CTX *int_ctx_new(OSSL_LIB_CTX *libctx,
             }
 #ifndef FIPS_MODULE
             if (keytype != NULL) {
+                OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CONFIG, NULL);
```

Comment:
Why is this associated with `OPENSSL_INIT_LOAD_CONFIG`????? This seems to have very little to do with the config file?

Reply:
In `EVP_PKEY_CTX_new_id` we call `OPENSSL_init_crypto` so we are sort of consistent here. And it's worth loading config anyway - it may contain the proper provider.

Reply:
> And it's worth loading config anyway

No!! Really don't do this. This way leads to anarchy and deadlocks. Loading the config file is a heavy operation that itself may end up calling `OPENSSL_init_crypto`. We've had to be very careful about it to avoid these problems. It should *only* be used just immediately before where we are about to use a variable that can be impacted by the config file.

Reply:
Well... What's the difference with the current code loading config on the engine branch?

Reply:
> Well... What's the difference with the current code loading config on the engine branch?

I don't know where you are referring to?

Reply:
If we go through EVP_PKEY_CTX_new_id, we will first go through engine branch, that will call the OPENSSL_init_crypto

Reply:
> If we go through EVP_PKEY_CTX_new_id, we will first go through engine branch, that will call the OPENSSL_init_crypto

Where *specifically* do you see config being loaded on the "engine branch". Do you mean in `ossl_engine_table_select`?

Reply:
It is loaded by the OBJ_nid2sn() call https://github.com/openssl/openssl/blob/8f8b861bc437f86df08162d6f564ea52536aa6cb/crypto/evp/pmeth_lib.c#L234

If you look at the obj_dat.c it always implicitly initializes the openssl with `OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CONFIG,NULL)`

Reply:
Probably yes. Matt, sorry, I am not by the computer just now. If you call EVP_PKEY_CTX_new_id from the test I added with openssl built with, say, no-sm2 and NID_sm2 as an argument, you will get to OPENSSL_init_crypto

Reply:
IMO this call here could be replaced with ossl_init_casecmp() call. That should not have any bad side effects.

Reply:
> If you look at the obj_dat.c it always implicitly initializes the openssl with OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CONFIG,NULL)

Right. In that case the global variable `added` is impacted by loading the config file - so we need to make sure we've loaded it before attempting to read the value of the `added` variable. Everywhere we have such a global variable in the source that could be impacted by the config file, we make sure we load it via  `OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CONFIG,NULL)` - but we don't do so anywhere else (or at least we shouldn't). It just isn't necessary. As long as those variables are initialised before first use everything should work correctly. Littering the code with `OPENSSL_INIT_LOAD_CONFIG` where it is not needed is a bad idea and it has caused difficult to debug problems in the past. We should minimise our usage of this.

Reply:
Yes, I agree with you for master

Reply:
If it helps, I can make the change to `ossl_init_casecmp()` and open a new PR with it?

Reply:
If you do it, please copy the tests from this PR

Reply:
See #18175.